Autopilot/Src/IMU.cpp

```diff
@@ -0,0 +1,225 @@
+      write_data(writeData); 
+   }
+
+   accelConversionFactor = ACCEL_SENSITIVITY_4G;
+   gyroConversionFactor = GRYO_SENSITIVITY_500;
+   tempConversionFactor = TEMPERATURE_SENSITIVITY;
+}
+
+void ICM20602::write_data(uint8_t *writeData) { 
+   uint8_t imuPlaceholderByte[2] = {0x00, 0x00};                                                                            
+   spi_port->set_slave(imuSlaveIdentifier);
+   spi_port->exchange_data(writeData, imuPlaceholderByte, 2);                                                                  
+}
+
+void ICM20602::get_accel_temp_gyro_reading(float *accx, float *accy, float *accz, float *gyrx, float *gyry, float *gyrz, float *temp) {
+   uint8_t imuBufferRead_FourteenBytes[15];
+   imuBufferRead_FourteenBytes[0] = REG_ACCEL_XOUT_H | 0x80;
+
+   uint8_t imu_raw_data[15];
+   imu_raw_data[0] = 0x00; //Dummy byte for MOSI
+   
+   //Store both High and Low Byte values
+   int16_t shiftedSensorAccX, shiftedSensorAccY, shiftedSensorAccZ, shiftedSensorTemp, shiftedSensorGyroX, shiftedSensorGyroY, shiftedSensorGyroZ; 
+   
+   for(int i = 1; i < 15; i++) {  
+      //0x00 is the reset value for all registers (Page 31). Thus, 0xFF will be used for the sensor_read (read only) registers
+      imuBufferRead_FourteenBytes[i] = 0xFF; //Initializes all dummy bytes in the MISO  
+      imu_raw_data[i] = 0x00; 
+   }
+   
+   //Uses the sensor registers to get raw data for all sensors
+   spi_port->set_slave(imuSlaveIdentifier);
+   sensorSuccess = spi_port->exchange_data(imuBufferRead_FourteenBytes, imu_raw_data, 15); //Sending in 15 bytes to tx_data (One register byte and fourteen dummy bytes)
+   
+   shiftedSensorAccX = (imu_raw_data[1] << 8) + imu_raw_data[2];
+   shiftedSensorAccY = (imu_raw_data[3] << 8) + imu_raw_data[4];
+   shiftedSensorAccZ = (imu_raw_data[5] << 8) + imu_raw_data[6];
+
+   shiftedSensorTemp = (imu_raw_data[7] << 8) + imu_raw_data[8];
+
+   shiftedSensorGyroX = (imu_raw_data[9] << 8) + imu_raw_data[10];
+   shiftedSensorGyroY = (imu_raw_data[11] << 8) + imu_raw_data[12];
+   shiftedSensorGyroZ = (imu_raw_data[13] << 8) + imu_raw_data[14];
+
+   //Converts 16-bit integers to a float. These are the actual measurements
+   *accx = ((float) shiftedSensorAccX)/accelConversionFactor; 
+   *accy = ((float) shiftedSensorAccY)/accelConversionFactor; 
+   *accz = ((float) shiftedSensorAccZ)/accelConversionFactor; 
+
+   *temp = ((float) shiftedSensorTemp)/tempConversionFactor;
+
+   *gyrx = ((float) shiftedSensorGyroX)/gyroConversionFactor;
+   *gyry = ((float) shiftedSensorGyroY)/gyroConversionFactor;
+   *gyrz = ((float) shiftedSensorGyroZ)/gyroConversionFactor;
+}
+ 
+void ICM20602::Begin_Measuring() {
+   get_accel_temp_gyro_reading(&measuredAccX, &measuredAccY, &measuredAccZ, &measuredGyroX, &measuredGyroY, &measuredGyroZ, &measuredTemp);
+   dataIsNew = true;
+}
+ 
+void ICM20602::GetResult(IMUData_t &Data) {
+   dataIsNew = false;
+   if (sensorSuccess == STATUS_CODE_OK) {
```

Comment:
there's a bug here. if I call GetResult twice in a row, it will tell me data is new both times rather than telling me it's old the second time.